cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,209 @@
+
+    raft::copy(frontier_vertices.begin() + starting_vertices.size(),
+               new_frontier_vertices.data(),
+               new_frontier_vertices.size(),
+               handle.get_stream());
+  } else {
+    raft::copy(frontier_vertices.begin(),
+               new_frontier_vertices.data(),
+               new_frontier_vertices.size(),
+               handle.get_stream());
+  }
+
+  if (frontier_vertex_labels) {
+    if (carry_over_sources) {
+      raft::copy(frontier_vertex_labels->begin(),
+                 starting_vertex_labels->data(),
+                 starting_vertex_labels->size(),
+                 handle.get_stream());
+
+      raft::copy(frontier_vertex_labels->begin() + starting_vertices.size(),
+                 new_frontier_vertex_labels->data(),
+                 new_frontier_vertex_labels->size(),
+                 handle.get_stream());
+    } else {
+      raft::copy(frontier_vertex_labels->begin(),
+                 new_frontier_vertex_labels->data(),
+                 new_frontier_vertex_labels->size(),
+                 handle.get_stream());
+    }
+  }
+
+  if constexpr (multi_gpu) {
+    if (frontier_vertex_labels) {
+      std::tie(frontier_vertices, *frontier_vertex_labels) =
+        shuffle_int_vertex_value_pairs_to_local_gpu_by_vertex_partitioning(
+          handle,
+          std::move(frontier_vertices),
+          std::move(*frontier_vertex_labels),
+          vertex_partition_range_lasts);
+    } else {
+      frontier_vertices = shuffle_int_vertices_to_local_gpu_by_vertex_partitioning(
+        handle, std::move(frontier_vertices), vertex_partition_range_lasts);
+    }
+  }
+
+  if (vertex_used_as_source) {
+    if (frontier_vertex_labels) {
+      std::tie(frontier_vertices, *frontier_vertex_labels) =
+        remove_visited_vertices_from_frontier(handle,
+                                              std::move(frontier_vertices),
+                                              std::move(*frontier_vertex_labels),
+                                              std::get<0>(*vertex_used_as_source),
+                                              *std::get<1>(*vertex_used_as_source));
+    } else {
+      auto new_end =
+        thrust::copy_if(handle.get_thrust_policy(),
+                        frontier_vertices.begin(),
+                        frontier_vertices.end(),
+                        frontier_vertices.begin(),
+                        [d_vertex_used = std::get<0>(*vertex_used_as_source).data(),
+                         d_vertex_size = std::get<0>(*vertex_used_as_source).size(),
+                         d_vertex_partition] __device__(vertex_t v) {
+                          return !thrust::binary_search(
+                            thrust::seq, d_vertex_used, d_vertex_used + d_vertex_size, v);
+                        });
+      frontier_vertices.resize(thrust::distance(frontier_vertices.begin(), new_end),
+                               handle.get_stream());
+    }
+  }
```

Comment:
What's the partitioning requirements for this & next frontier vertices (vertex & label pairs)? And this code is quite confusing. We had `if (vertex_used_as_source)` at the beginning and here again. Why are we doing this? We really need to properly document the behavior for different flag combinations.

Reply:
And shouldn't we better handle both the if and else cases in `remove_visited_vertices_from_frontier`?

Reply:
The first `vertex_used_as_source` block was to maintain the list (hopefully I'll reorganize that).  This block is actually using `vertex_used_as_source` to do the filtering.

`this_frontier_vertices` (and the corresponding labels) are partitioned onto the GPU responsible for the vertex id.

`next_frontier_vertices` (and the corresponding labels) are not partitioned properly, the span includes destination vertices encountered on this GPU.  The shuffling step above gets the vertices shuffled to the proper location.

I will modify `remove_visited_vertices_from_frontier` to handle both cases.

Reply:
Reordered code in this function to only use `vertex_used_as_source` once which should reduce confusion.

Modified to use `set_union` in next push

Reply:
OK... reverted from the `set_union` approach after testing.

`set_union` has some preconditions - notably that both inputs must be sorted and unique.  Because one of the inputs to this is a device span pointing to an immutable object that is neither sorted nor unique, I need to make a copy of that data.  I don't think the `set_union` approach ends up being any better than the new structure.

I did restructure the code and renamed some variables to make things a little clearer.  The restructured code should be easier to follow, and includes making sure that the data is shuffled onto the proper GPU prior to doing the sort/unique and using the values for filtering.